runtime/criusupport/criusupport.cpp

```diff
@@ -209,6 +209,63 @@ toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend)
 	}
 }
 
+static UDATA
+notCheckpointSafeFrameWalkFunction(J9VMThread *vmThread, J9StackWalkState *walkState)
+{
+	J9Method *method = walkState->method;
+	if (NULL != method) {
+		J9ClassLoader *methodLoader = J9_CLASS_FROM_METHOD(method)->classLoader;
+		/* we only enforce this in methods loaded by the bootloader */
+		if ((NULL == methodLoader) || (methodLoader == vmThread->javaVM->systemClassLoader)) {
+			J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+			if (NULL != romMethod && J9ROMMETHOD_HAS_EXTENDED_MODIFIERS(romMethod)) {
+				U_32 extraModifiers = getExtendedModifiersDataFromROMMethod(romMethod);
+				if (J9ROMMETHOD_HAS_NOT_CHECKPOINT_SAFE_ANNOTATION(extraModifiers)) {
+					*(bool *)walkState->userData1 = false;
+					walkState->userData2 = (void *)vmThread;
+					walkState->userData3 = (void *)method;
+					return J9_STACKWALK_STOP_ITERATING;
+				}
+			}
+		}
+	}
+
+	return J9_STACKWALK_KEEP_ITERATING;
+}
+
+static bool
+checkIfSafeToCheckpoint(J9VMThread *currentThread)
+{
+	bool isSafe = true;
+	J9JavaVM *vm = currentThread->javaVM;
+
+	Assert_CRIU_true((J9_XACCESS_EXCLUSIVE == vm->exclusiveAccessState) || (J9_XACCESS_EXCLUSIVE == vm->safePointState));
+
+	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
+	while (NULL != walkThread) {
+		if (VM_VMHelpers::threadCanRunJavaCode(walkThread)
+		&& (currentThread != walkThread)
+		) {
+			isSafe = true;
+			J9StackWalkState walkState;
+			walkState.walkThread = walkThread;
+			walkState.flags = J9_STACKWALK_ITERATE_FRAMES;
+			walkState.skipCount = 0;
+			walkState.userData1 = (void *)&isSafe;
+			walkState.frameWalkFunction = notCheckpointSafeFrameWalkFunction;
+
+			vm->walkStackFrames(walkThread, &walkState);
+			if (false == *(bool *)walkState.userData1) {
```

Comment:
IsSafe is never set to false. Probably want to set it to true in the declaration and not set it at the top of the loop.

Here, set isSafe to userData1 and break if false.

Reply:
I dont follow `isSafe` is `userData1`

```
walkState.userData1 = (void *)&isSafe;
```

Reply:
Sorry, miseed the fact you are passing in a pointer to the variable. This line should check the variable directly, not the userData.